### PR TITLE
MAID-3275: Reduce PublicId lookups

### PR DESCRIPTION
**Pull request by madadam**
This PR replaces `PublicId` with `PeerIndex` in almost all places. As `PeerIndex` is just a newtype over `usize`, this reduces the time spend looking up entries in maps end sets keyed by `PeerIndex`. Further, it replaces all `BTreeMap` / `BTreeSet` of small keys (`EventIndex` or `PeerIndex`) with `FnvHashMap` and `FnvHashSet` from the [fnv crate](https://github.com/servo/rust-fnv), which further speeds the lookups up.

As `PeerIndex` is no longer "objective" (different peers will have different `PeerIndex` identifying the same peer), this PR also modifies the test code to account for that (similarly to how we handled replacing `EventHash` with `EventIndex` in the past).

### CODE REVIEW

File: src/peer_list/peer.rs [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+            .events
+            .indexed()
+            .rev()
+            .next()
+            .map(|(index, _)| index)
+            .unwrap_or(0);
+        self.membership_list_changes.push((index, change));
+    }
+
+    pub(crate) fn membership_list(&self) -> &BTreeSet<P> {
```
Comment (jeanphilippeD) [inline src/peer_list/peer.rs]: Do we need to remove the #[cfg(feature = "dump-graphs")], the member is public for the module that needs it?

File: src/peer_list/peer.rs [hunk 2]
```diff
@@ -85,17 +93,51 @@ impl<P: PublicId> Peer<P> {
...
+impl Peer<PeerId> {
+    pub(super) fn new_from_dot_input(
+        peer_id: &PeerId,
+        state: PeerState,
+        membership_list: BTreeSet<PeerId>,
+        graph: &Graph<Transaction, PeerId>,
+    ) -> Self {
+        let mut events = Events::new();
+        for event in graph {
+            if event.creator() == peer_id {
```
Comment (jeanphilippeD) [inline src/peer_list/peer.rs]: This was dropped, why?

```
} else if !peer_ids.contains(event.creator()) {
                    debug!(
                        "peer_data list doesn't contain the creator of event {:?}",
                        *event
                    );
                }
```

File: src/peer_list/mod.rs [hunk 3]
```diff
@@ -435,35 +433,12 @@ impl PeerList<PeerId> {
...
-                .into_iter()
-                .chain(iter::once(peer_id.clone()))
-                .map(MembershipListChange::Add)
-            {
-                peer.change_membership_list(change)
-            }
-
-            let _ = peers.insert(peer_id, peer);
-        }
+        let peers = peer_data
```
Comment (jeanphilippeD) [inline src/peer_list/mod.rs]: This is so much better.

File: src/dump_graph.rs [hunk 4]
```diff
@@ -116,15 +116,16 @@ mod detail {
     thread_local!(static DUMP_COUNTS: RefCell<BTreeMap<String, usize>> =
         RefCell::new(BTreeMap::new()));
 
-    fn catch_dump<T: NetworkEvent, P: PublicId>(
+    fn catch_dump<T: NetworkEvent, S: SecretId>(
```
Comment (jeanphilippeD) [inline src/dump_graph.rs]: So there is a lot of changes in this commit 'chore/peer_list: use PeerIndex instead of PublicId ' that seem to be related to testing and dump graph. I assume our automated tests / unit test will cover the parsing, Does it cover the dumpgraph, what kind of test for that?

File: src/parsec.rs [hunk 5]
```diff
@@ -1588,7 +1584,7 @@ impl<T: NetworkEvent, S: SecretId> Parsec<T, S> {
 
         // IMPORTANT: We must sort this in consistent order, so when the tie breaking rule kicks in,
         // the outcome is the same for everyone.
-        payloads.sort();
+        payloads.sort_by(|a, b| a.hash().cmp(b.hash()));
```
Comment (jeanphilippeD) [inline src/parsec.rs]: Not clear to me why this change?

File: src/dev_utils/dot_parser.rs [hunk 6]
```diff
@@ -14,7 +14,7 @@ use meta_voting::{
 };
 use mock::{PeerId, Transaction};
 use observation::{Observation, ObservationHash, ObservationKey};
-use peer_list::{PeerIndex, PeerList, PeerState};
+use peer_list::{PeerIndexMap, PeerIndexSet, PeerList, PeerState};
```
Comment (jeanphilippeD) [inline src/dev_utils/dot_parser.rs]: That is great, should really help when trying different maps.

File: src/dev_utils/dot_parser.rs [hunk 7]
```diff
@@ -688,7 +688,7 @@ impl ParsedContents {
     /// Create empty `ParsedContents`.
     pub fn new(our_id: PeerId) -> Self {
         let peer_list = PeerList::new(our_id.clone());
-        let meta_elections = MetaElections::new(BTreeSet::new());
+        let meta_elections = MetaElections::new(PeerIndexSet::default());
```
Comment (jeanphilippeD) [inline src/dev_utils/dot_parser.rs]: should this be in previous commit?
I think it most likely should.
Same of similar updates throughout in this commit.

File: src/functional_tests.rs [hunk 8]
```diff
@@ -222,23 +222,29 @@ fn add_peer() {
...
     let event_index = 18;
-    let mut alice_membership_list_for_dave = vec![
+    let mut expected_alice_membership_list_for_dave = vec![
         (event_index, MembershipListChange::Add(alice_index)),
         (event_index, MembershipListChange::Add(bob_index)),
         (event_index, MembershipListChange::Add(carol_index)),
         (event_index, MembershipListChange::Add(dave_index)),
         (event_index, MembershipListChange::Add(eric_index)),
     ];
+    expected_alice_membership_list_for_dave.sort();
```
Comment (jeanphilippeD) [inline src/functional_tests.rs]: Why the sort?

File: src/meta_voting/meta_elections.rs [hunk 9]
```diff
@@ -42,7 +43,7 @@ impl Debug for MetaElectionHandle {
 
 #[derive(Debug, Clone, PartialEq, Eq)]
 pub(crate) struct MetaElection {
-    pub(crate) meta_events: BTreeMap<EventIndex, MetaEvent>,
+    pub(crate) meta_events: FnvHashMap<EventIndex, MetaEvent>,
```
Comment (jeanphilippeD) [inline src/meta_voting/meta_elections.rs]: Would it make sense to have the same kind of type alias as for PeerIndexMap, PeerIndexSet

File: src/peer_list/peer.rs (hunk 1 above)
Comment (madadam) [reply src/peer_list/peer.rs]: If I understand correctly what you are asking, the yes, the attribute had to go because this function is now needed elsewhere, not just in `dump_graph`, because the `membership_list` field is no longer accessible outside of this module.

File: src/peer_list/peer.rs (hunk 2 above)
Comment (madadam) [reply src/peer_list/peer.rs]: We no longer have access to the necessary data to perform the check here. It was just additional sanity check to make sure all the events are created by someone we know. Don't think the check is essential. I can put it back to some more appropriate place if you thinks it's worth it.

File: src/dump_graph.rs (hunk 4 above)
Comment (madadam) [reply src/dump_graph.rs]: Yes, the `dev_utils::dot_parser::tests::dot_parser` also covers dumping. Also, as mentioned in other comment, the changes were necessary to make everything still compile. Even though they might seem unrelated to the purpose of this commit.

File: src/parsec.rs (hunk 5 above)
Comment (madadam) [reply src/parsec.rs]: As the comment says, we need consistent order here (every node should arrive at the same order), because this is the point where we pick the next consensused block. `payloads` contain `ObservationKey` which contains `PeerIndex`, but `PeerIndex` is not consistent (different nodes might use different `PeerIndex` to refer to the same peer), so we need to exclude it from the comparison.

File: src/dev_utils/dot_parser.rs (hunk 7 above)
Comment (madadam) [reply src/dev_utils/dot_parser.rs]: If we want to be pedantic then yes, it should. Problem is the compiler didn't complain that I'm using `BTreeSet` instead of `PeerIndexSet` because it was just a type alias at this point. When I changed it to `FnvHashSet` then it started to complain, but it was already in different commit and I didn't feel like going back and changing the previous commit.

File: src/functional_tests.rs (hunk 8 above)
Comment (madadam) [reply src/functional_tests.rs]: The values now come out of `HashMap` which doesn't guarantee any order. Besides, for the same `event_index`, the order isn't important anyway.

File: src/meta_voting/meta_elections.rs (hunk 9 above)
Comment (madadam) [reply src/meta_voting/meta_elections.rs]: It might, but this isn't used nearly as much as `PeerIndexSet` / `PeerIndexMap` so I didn't feel the need.

File: src/peer_list/peer.rs (hunk 2 above)
Comment (jeanphilippeD) [reply src/peer_list/peer.rs]: No it's ok. Would have been good to have mentioned it in commit comment as it is lost in the diff.

File: src/dump_graph.rs (hunk 4 above)
Comment (jeanphilippeD) [reply src/dump_graph.rs]: I guess my question was did you test/spot check that the dump graph functionality still work (beside it compiling), or is the automated test that did that for you?

Comment (madadam) [reply src/dump_graph.rs]: I relied on the tests only. The way it works is that it generates some parsec graphs using the testing framework, dumps them, then parses them back and compares with the originals. Seems thorough enough so I didn't bother checking it manually.

File: src/parsec.rs [hunk 10]
```diff
@@ -479,9 +487,13 @@ impl<T: NetworkEvent, S: SecretId> Parsec<T, S> {
         }
     }
 
+    fn get_peer_index(&self, peer_id: &S::PublicId) -> Result<PeerIndex> {
```
Comment (Fraser999) [inline src/parsec.rs]: Getters normally shouldn't be prefixed with `get_`.

File: src/peer_list/membership_list.rs [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+impl MembershipListChange {
+    pub(super) fn apply(&self, peers: &mut PeerIndexSet) -> bool {
+        match *self {
+            MembershipListChange::Add(index) => peers.insert(index),
+            MembershipListChange::Remove(index) => peers.remove(&index),
+        }
+    }
+
+    #[cfg(feature = "malice-detection")]
+    pub(super) fn unapply(&self, peers: &mut PeerIndexSet) -> bool {
```
Comment (Fraser999) [inline src/peer_list/membership_list.rs]: "unapply" could be "revert" maybe?

File: src/parsec.rs (hunk 10 above)
Comment (madadam) [reply src/parsec.rs]: This isn't a getter though. I think getter would be something that doesn't take any arguments, no?

Comment (Fraser999) [reply src/parsec.rs]: Not sure.  Maybe that is the strict definition, but I guess the spirit of that convention would still apply here, wouldn't it?  For example, the next function could have been called `get_our_last_event_hash`?

Not a big deal anyway - it's purely a style issue.  I'm fine to approve as is.
